src/domain/mongoose_domain_loader.erl

```diff
@@ -1,13 +1,58 @@
+%% This module is designed assuming the fact, that records inserted
+%% into domains or events table could appear in any order.
+%% I.e. events with ids [1, 2, 3] could appear as [1,3] for a short amount of time.
+%% We also assume, event ids are never reused.
 -module(mongoose_domain_loader).
--export([load_data_from_base/2,
-         check_for_updates/2,
-         remove_outdated_domains_from_core/0]).
+-export([initial_load/0,
+         check_for_updates/0]).
+
+%% For tests
+-export([find_gaps_between/2]).
 
 -include("mongoose_logger.hrl").
 
+%% There are two important functions, called by service_domain_db:
+%% - initial_load
+%% - check_for_updates
+initial_load() ->
+    case mongoose_loader_state:get(undefined) of
+        undefined ->
+            %% If mongoose_loader_state is undefined,
+            %% this means we start for the first time with this core process
+            cold_load();
+        reset ->
+            %% Case when state has been reset without restarting core
+            %% For example, when we detected out-of-sync situation
+            cold_load(),
+            remove_outdated_domains_from_core();
+        _ ->
+            %% Already synced to some point.
+            %% Just read updates from the event table.
+            skip
+    end.
+
+%% Load from the domain table
+cold_load() ->
+    %% We assume that to sync successfully we need conditions:
+    %% - events table does not contain gaps (if it contain a gap, a record could be missing).
+    %% - we have to check the whole event table for gaps.
+    %% - we don't care about gaps in the domain table - even if some state
+    %%   is not visible yet in the domain table, it would be visible once
+    %%   we try to fix the event gap.
+    {MinEventId, MaxEventId} = mongoose_domain_sql:get_minmax_event_id(),
+    %% It's important to get gaps info before the loading of domains
+    Gaps = find_gaps_between(MinEventId, MaxEventId),
+    %% Do domain loading from the main domain table
+    load_data_from_base(0, 1000),
+    %% Try to fix gaps
```

Comment:
Why do we fix the gaps? What is the problem with leaving the gaps as they are? How would a different node in this cluster know that there was a gap if we fix it before it gets filled by the DB with a real event?

Reply:
We fix gaps, so we don't need to track and query for them, once we finally read a record with the gap's event id. Because a gap means the data could appear there. 
INSERTs are never overwritten. Only one transaction would pass, another would get aborted.
Another node would get the committed data (any node actually would only get committed data).

Reply:
So if we fix gaps on one node in the cluster while on some node (the same or another) we try to insert a domain, the insert might fail, right? Is this really rare as we experienced gaps in tests and load tests quite frequently?

Reply:
it is pretty rare, usually events appear in the order of the incremented keys (i.e. without gaps).
But even if we see the gap and trying to fix it, by the time we apply the fix, that fix gets rejected (because the actual data appears).

Reply:
if we insert 3000 events **at once**, there could be 1-5 gaps. But all events would get inserted properly, just after a retry.
But I am pretty sure it is a very edge case for the actual users, the actual use case would be probably a couple of domains inserted per minute.

Reply:
Ok, so we have retries for domain insertion, that's good.